apollo-spaceport/src/server.rs

```diff
@@ -0,0 +1,330 @@
+        }
+    }
+}
+
+/// Accept Traces and Stats from clients and transfer to an Apollo Ingress
+pub struct ReportSpaceport {
+    addr: SocketAddr,
+    // This Map will only have a single entry if used internally from a router.
+    // (because a router can only be serving a single graph)
+    graph_usage: GraphUsageMap,
+    tx: Sender<()>,
+    total: Arc<AtomicU32>,
+}
+
+impl ReportSpaceport {
+    /// Create a new ReportSpaceport which is configured to serve requests at the
+    /// supplied address
+    ///
+    /// The spaceport will buffer data and attempt to transfer it to the Apollo Ingress
+    /// every 5 seconds. This transfer will be triggered sooner if data is
+    /// accumulating more quickly than usual.
+    ///
+    /// The spaceport will attempt to make the transfer 5 times before failing. If
+    /// the spaceport fails, the data is discarded.
+    pub fn new(addr: SocketAddr) -> Self {
+        // Spawn a task which will check if there are reports to
+        // submit every interval.
+        let graph_usage: GraphUsageMap = Arc::new(Mutex::new(HashMap::new()));
+        let task_graph_usage = graph_usage.clone();
+        let (tx, mut rx) = tokio::sync::mpsc::channel::<()>(10);
+        let total = Arc::new(AtomicU32::new(0u32));
+        let task_total = total.clone();
+        tokio::task::spawn(async move {
+            let client = Client::new();
+            let mut interval = interval(INGRESS_CLOCK_TICK);
+            interval.set_missed_tick_behavior(MissedTickBehavior::Skip);
+            interval.tick().await;
+            loop {
+                tokio::select! {
+                    biased;
+                    mopt = rx.recv() => {
+                        tracing::debug!("spaceport triggered");
+                        match mopt {
+                            Some(_msg) => {
+                                for result in extract_graph_usage(&client, task_graph_usage.clone()).await {
+                                    match result {
+                                        Ok(v) => tracing::debug!("Report submission succeeded: {:?}", v),
+                                        Err(e) => tracing::error!("Report submission failed: {}", e),
+                                    }
+                                }
+                            },
+                            None => break
+                        }
+                    },
+                    _ = interval.tick() => {
+                        tracing::debug!("spaceport ticked");
+                        task_total.store(0, Ordering::SeqCst);
+                        for result in extract_graph_usage(&client, task_graph_usage.clone()).await {
+                            match result {
+                                Ok(v) => tracing::debug!("Report submission succeeded: {:?}", v),
+                                Err(e) => tracing::error!("Report submission failed: {}", e),
+                            }
+                        }
```

Comment:
Code's a bit wide here and this is the same `for` above, so was kinda wondering if this should just be absorbed into a (renamed) `extract_graph_usage`.

Reply:
I added a wrapper function which contains the loop: process_all_graphs.
done: d6fb85ecc376a0c4ff60ebe20f5cf657d64a5659

Reply:
thanks! looks great!